### PR TITLE
Adds a PointsProcessor that can write ROS maps.

### LINKED ISSUE #475: Generating pgm/yaml map
Since the output of the 'finish_trajectory' service has changed, I'm unable to generate .pgm/.yaml map files for other nodes (non-Cartographer) to use.

I'm aware of the 'write_state' service which provides a .pbstream file, but is it still possible to write to a .pgm/.yaml file?

### PR DESCRIPTION
**Pull request by SirVer**
Fixes #475.

Depends on googlecartographer/cartographer#598.

### CODE REVIEW

**Comment by SirVer** [comment]
@gaschler ping?

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// out a PGM and YAML suitable for ROS map server to consume.
+class RosMapWritingPointsProcessor
+    : public ::cartographer::io::PointsProcessor {
+ public:
+  constexpr static const char* kConfigurationFileActionName = "write_ros_map";
+  RosMapWritingPointsProcessor(
+      double resolution,
+      const ::cartographer::mapping_2d::proto::RangeDataInserterOptions&
+          range_data_inserter_options,
+      ::cartographer::io::FileWriterFactory file_factory,
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h]: file_writer_factory

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.cc [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  const double resolution = limits.resolution();
+  const double x_offset =
+      limits.max().x() - (offset.y() + image.height()) * resolution;
+  const double y_offset =
+      limits.max().y() - (offset.x() + image.width()) * resolution;
+  const std::string output =
+      "image: " + pgm_filename + "\n" +
+      "resolution: " + std::to_string(resolution) + "\n" + "origin: [" +
+      std::to_string(x_offset) + ", " + std::to_string(y_offset) +
+      ", 0.]\nnegate: 0\noccupied_thresh: 0.65\nfree_thresh: 0.196\n";
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/ros_map_writing_points_processor.cc]: Define these magic numbers as constants?

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      const string& filestem, PointsProcessor* next);
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h]: PointsProcesser* const?

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.cc (hunk 2 above)
Comment (SirVer) [reply cartographer_ros/cartographer_ros/ros_map_writing_points_processor.cc]: I added a link where they come from.

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h (hunk 3 above)
Comment (SirVer) [reply cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h]: Not in the header - it has no meaning for the API, it is purely an implementation detail.

File: cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      ::cartographer::common::LuaParameterDictionary* dictionary,
+      PointsProcessor* next);
+
+  ~RosMapWritingPointsProcessor() override {}
+
+  void Process(std::unique_ptr<::cartographer::io::PointsBatch> batch) override;
+  FlushResult Flush() override;
+
+ private:
+  const string filestem_;
```
[the remark this answers is not in the document]
Comment (SirVer) [reply cartographer_ros/cartographer_ros/ros_map_writing_points_processor.h]: Stem is the basename without the extension, i.e. for /foo/bar/blub.png stem is blub while basename would be blub.png.

**Pull request merged**